// fnbr/dist/src/structures/ClientUser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const User_1 = (0, tslib_1.__importDefault)(require("./User"));
/**
 * Represents the user of a client
 */
class ClientUser extends User_1.default {
    /**
     * @param client The main client
     * @param data The user's data
     */
    constructor(client, data) {
        super(client, data);
        this.name = data.name;
        this.lastName = data.lastName;
        this.email = data.email;
        this.failedLoginAttempts = data.failedLoginAttempts;
        this.lastLogin = new Date(data.lastLogin);
        this.numberOfDisplayNameChanges = data.numberOfDisplayNameChanges;
        this.ageGroup = data.ageGroup;
        this.headless = data.headless;
        this.country = data.country;
        this.preferredLanguage = data.preferredLanguage;
        this.canUpdateDisplayName = data.canUpdateDisplayName;
        this.tfaEnabled = data.tfaEnabled;
        this.emailVerified = data.emailVerified;
        this.minorVerified = data.minorVerified;
        this.minorExpected = data.minorExpected;
        this.minorStatus = data.minorStatus;
    }
}
exports.default = ClientUser;
