// fnbr/dist/src/structures/PartyMemberMeta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Meta_1 = (0, tslib_1.__importDefault)(require("../util/Meta"));
/**
 * Represents a party member meta
 */
class PartyMemberMeta extends Meta_1.default {
    /**
     * @param member The party member
     * @param schema The schema
     */
    constructor(member, schema) {
        super(schema);
        this.member = member;
    }
    /**
     * The currently equipped outfit CID
     */
    get outfit() {
        var _a, _b, _c, _d;
        return (_d = (_c = (_b = (_a = this.get('Default:AthenaCosmeticLoadout_j')) === null || _a === void 0 ? void 0 : _a.AthenaCosmeticLoadout) === null || _b === void 0 ? void 0 : _b.characterDef) === null || _c === void 0 ? void 0 : _c.match(/(?<=\w*\.)\w*/)) === null || _d === void 0 ? void 0 : _d.shift();
    }
    /**
     * The currently equipped pickaxe ID
     */
    get pickaxe() {
        var _a, _b, _c, _d;
        return (_d = (_c = (_b = (_a = this.get('Default:AthenaCosmeticLoadout_j')) === null || _a === void 0 ? void 0 : _a.AthenaCosmeticLoadout) === null || _b === void 0 ? void 0 : _b.pickaxeDef) === null || _c === void 0 ? void 0 : _c.match(/(?<=\w*\.)\w*/)) === null || _d === void 0 ? void 0 : _d.shift();
    }
    /**
     * The current emote EID
     */
    get emote() {
        var _a, _b, _c;
        const emoteAsset = (_b = (_a = this.get('Default:FrontendEmote_j')) === null || _a === void 0 ? void 0 : _a.FrontendEmote) === null || _b === void 0 ? void 0 : _b.emoteItemDef;
        if (emoteAsset === 'None' || !emoteAsset)
            return undefined;
        return (_c = emoteAsset.match(/(?<=\w*\.)\w*/)) === null || _c === void 0 ? void 0 : _c.shift();
    }
    /**
     * The currently equipped backpack BID
     */
    get backpack() {
        var _a, _b, _c, _d;
        return (_d = (_c = (_b = (_a = this.get('Default:AthenaCosmeticLoadout_j')) === null || _a === void 0 ? void 0 : _a.AthenaCosmeticLoadout) === null || _b === void 0 ? void 0 : _b.backpackDef) === null || _c === void 0 ? void 0 : _c.match(/(?<=\w*\.)\w*/)) === null || _d === void 0 ? void 0 : _d.shift();
    }
    /**
     * Whether the member is ready
     */
    get isReady() {
        var _a, _b;
        return ((_b = (_a = this.get('Default:LobbyState_j')) === null || _a === void 0 ? void 0 : _a.LobbyState) === null || _b === void 0 ? void 0 : _b.gameReadiness) === 'Ready';
    }
    /**
     * The current input method
     */
    get input() {
        return this.get('Default:CurrentInputType_s');
    }
    /**
     * The cosmetic variants
     */
    get variants() {
        var _a;
        const variants = (_a = this.get('Default:AthenaCosmeticLoadoutVariants_j')) === null || _a === void 0 ? void 0 : _a.vL;
        if (!variants)
            return {};
        const pascalCaseVariants = {};
        Object.keys(variants).forEach((k) => {
            pascalCaseVariants[`${k.charAt(0).toUpperCase()}${k.slice(1)}`] = variants[k];
        });
        return pascalCaseVariants;
    }
    /**
     * The custom data store
     */
    get customDataStore() {
        var _a;
        return ((_a = this.get('Default:ArbitraryCustomDataStore_j')) === null || _a === void 0 ? void 0 : _a.ArbitraryCustomDataStore) || [];
    }
    /**
     * The banner info
     */
    get banner() {
        var _a;
        return (_a = this.get('Default:AthenaBannerInfo_j')) === null || _a === void 0 ? void 0 : _a.AthenaBannerInfo;
    }
    /**
     * The battle pass info
     */
    get battlepass() {
        var _a;
        return (_a = this.get('Default:BattlePassInfo_j')) === null || _a === void 0 ? void 0 : _a.BattlePassInfo;
    }
    /**
     * The platform
     */
    get platform() {
        var _a, _b, _c, _d;
        return (_d = (_c = (_b = (_a = this.get('Default:PlatformData_j')) === null || _a === void 0 ? void 0 : _a.PlatformData) === null || _b === void 0 ? void 0 : _b.platform) === null || _c === void 0 ? void 0 : _c.platformDescription) === null || _d === void 0 ? void 0 : _d.name;
    }
    /**
     * The match info
     */
    get match() {
        const location = this.get('Default:Location_s');
        const hasPreloadedAthena = this.get('Default:HasPreloadedAthena_b');
        const isSpectatable = this.get('Default:SpectateAPartyMemberAvailable_b');
        const playerCount = this.get('Default:NumAthenaPlayersLeft_U');
        const matchStartedAt = this.get('Default:UtcTimeStartedMatchAthena_s');
        return {
            hasPreloadedAthena,
            isSpectatable,
            location,
            matchStartedAt: matchStartedAt && new Date(matchStartedAt),
            playerCount,
        };
    }
    /**
     * Whether a marker has been set
     */
    get isMarkerSet() {
        var _a, _b;
        return !!((_b = (_a = this.get('Default:FrontEndMapMarker_j')) === null || _a === void 0 ? void 0 : _a.FrontEndMapMarker) === null || _b === void 0 ? void 0 : _b.bIsSet);
    }
    /**
     * The marker location [x, y] tuple. [0, 0] if there is no marker set
     */
    get markerLocation() {
        var _a, _b;
        const marker = (_b = (_a = this.get('Default:FrontEndMapMarker_j')) === null || _a === void 0 ? void 0 : _a.FrontEndMapMarker) === null || _b === void 0 ? void 0 : _b.markerLocation;
        if (!marker)
            return [0, 0];
        return [marker.y, marker.x];
    }
    /**
     * The assisted challenge
     */
    get assistedChallenge() {
        var _a;
        const challenge = (_a = this.get('Default:AssistedChallengeInfo_j')) === null || _a === void 0 ? void 0 : _a.AssistedChallengeInfo;
        if (!challenge)
            return undefined;
        return {
            questItemDef: challenge.questItemDef,
            objectivesCompleted: challenge.objectivesCompleted,
        };
    }
}
exports.default = PartyMemberMeta;
